[Experimental] Additional Hand Menu Layout Examples

alandergrouse opened the pull request:
## Overview
A new Hand Menu Example Layouts scene, with 4 layout options and tuning unique Hand Constraint Palm Up component.

![191107_HandMenuLayoutExamples_Screenshot](https://user-images.githubusercontent.com/38055323/68433039-7a67b700-016a-11ea-9a49-e837e30f2266.png)


## Changes
- additional hand menu layout prefabs with unique hand constraint tunings
- added toggle menu to switch between layouts
- customized slider prefab and script to one of the hand menu layouts
- edited scene description text
- polished toggle panel in original Hand Menu Example scene

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6503) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6503) before we can accept your contribution.<br/><br/>:x: alandergrouse [sign now](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6503)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/MixedRealityToolkit-Unity?pullRequest=6503) it.</sub>

wiwei commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/HandMenu_Slider_Assets/Scripts/UpdateTrackScale.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections;
```
Nit: We haven't been using underscores in folder names (i.e. HandMenu_Slider_Assets -> HandeMenuSliderAssets)

alandergrouse replied:
Thanks! Updating now.

wiwei commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/HandMenu_Slider_Assets/Scripts/UpdateTrackScale.cs:
```diff
@@ -0,0 +1,19 @@
+using Microsoft.MixedReality.Toolkit.UI;
+
+
+public class UpdateTrackScale : MonoBehaviour
```
This class needs to be placed into a namespace (any namespace, but ideally obviously a namespace that makes sense :))

Otherwise I think that this will trip the build system (i.e. some of our packaging parts makes sure that each file has a namespace)

alandergrouse replied:
I see! Would you be able to help me with that? I'm not too familiar with adding namespaces.

wiwei replied:
Ah, I had to look way too long for this - this is a bit interesting because the hand menu solver stuff (which is the existing code) lives in a namespace that this is on (i.e. solvers) that this code right here SHOULDN'T be in (because it's not a solver).

I'm thinking maybe something like:

Microsoft.MixedReality.Toolkit.Examples.Experimental.HandMenu

alandergrouse replied:
Thanks! I wrapped the script in Microsoft.MixedReality.Toolkit.Examples.Experimental.HandMenu namespace.

julenka replied:
Hi, @wiwei this is a pretty tiny nit but I think the guidance was to do more like `Microsoft.MixedReality.Toolkit.Experimental.Examples.HandMenu`. See [this page](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace) for guidance on namespaces for experimental features.

wiwei commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/HandMenu_Slider_Assets/Scripts/UpdateTrackScale.cs:
```diff
@@ -0,0 +1,19 @@
+using Microsoft.MixedReality.Toolkit.UI;
+
+
+public class UpdateTrackScale : MonoBehaviour
```
Is this intended to be an example script, or something that consumers will use?

If it's more of the latter, we'd probably want to expose roundedValue/trackScaleNode publicly as well so that these things are changeable at runtime by code.

alandergrouse replied:
The script is more just to make sure the active part of the track is updated based on the current level of the slider. I wouldn't say it should be customizable.

wiwei replied:
Gotcha.

I think that in this case we should have the fields commented (i.e

```csharp
// A value in the range [0, 100] indicating the slider position
... roundedValue...
```

alandergrouse replied:
Thanks! I just updated the script and removed the roundedValue altogether. It now references eventData.NewValue, which is more straightforward.

wiwei commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Scripts/UpdateSliderTrackLine.cs:
```diff
@@ -0,0 +1,17 @@
+    public void OnSliderUpdated(SliderEventData eventData)
+    {
+            roundedValue = Mathf.Round(eventData.NewValue * 100f) / 100f;
+            activeLine.transform.localScale = new Vector3 (transform.localScale.x, roundedValue, transform.localScale.z);
```
It feels a little weird to see two different classes which are basically completely the same except for the axis on which the change is applied - it feels like there should just be one MonoBehaviour with another dropdown/flag which controls whether or not the X or Y axis (or z axis!) is what is actually moved along.

alandergrouse replied:
I've taken out the rounded value and am just using eventData.NewValue to drive the y-axis scale of the track. Ideally this could be expanded to work for the X and Z axes, but I think that can happen in the next iteration.

davidkline-ms commented:
/azp run

davidkline-ms commented:
was running on an incorrect pipeline... requeuing

davidkline-ms commented:
/azp run

alandergrouse commented:
Thanks for the feedback, @julenka !

I'm reusing as many existing assets as possible, but as Unity doesn't support local material overrides, I've had to create a bunch of duplicate materials for new icons and the app tiles, for example. Those account for 54% of the newly added files.

In terms of additional file size, the total of the 111 added files is 1.42 MB:
![image](https://user-images.githubusercontent.com/38055323/68507263-5666ad00-0220-11ea-9ea7-2709cb968466.png)

The biggest portions of that are the new HandMenuLayoutExaploes scene (~700KB) and the prefabs (~400KB).

davidkline-ms commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/HandMenuSliderAssets/Scripts/UpdateTrackScale.cs:
```diff
@@ -0,0 +1,19 @@
+using Microsoft.MixedReality.Toolkit.UI;
+
+
+public class UpdateTrackScale : MonoBehaviour
```
this class still needs to be in a namespace.

alandergrouse replied:
Ended up removing this script as it was redundant. Thanks!

davidkline-ms commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/HandMenuSliderAssets/Scripts/UpdateTrackScale.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections;
```
Missing license header

davidkline-ms replied:
Here is the header to add:

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

Thanks!

davidkline-ms commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Scripts/UpdateSliderTrackLine.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections;
```
Missing license header.

alandergrouse replied:
Added!

alandergrouse commented:
Thanks for the feedback @Troy-Ferrell! Just pushed a commit with fixes for most of the issues you called out:

**Material instancing**
Is this something specific to the current scene, or the project in general? FWIW, only one hand menu is enabled at a time – I'm not sure if that makes a difference though. Is there something I can do to optimize, without impacting the design?

**Prefab fixes**
Good catch on on the missing parameters in the prefabs. There are some overrides in the scene that I've applied to the prefabs. This fixed the missing shaders in the 3x4 prefab as well as the HandConstraintPalmUp solvers.

**Overlap solvers**
The overlap solvers on the circular hand menu gives each button a different move lerp parameter, which makes for an "elastic" effect when the menu is moved along the hand. Here's a preview:
![GifMaker_20191108142358028](https://user-images.githubusercontent.com/38055323/68514778-8a979900-0233-11ea-8918-3035bbcc95fe.gif)


**Layout button selector**
One activates the hand menu, and the other ensures that repeated pressed of the toggle button don't turn off the BackPlateToggleState node. I hear you on the scalability problem. Ideally, we could drop in an array of game objects and run actions on them at once; I'm not sure how to go about that though. Perhaps we can edit the Interactable script?

alandergrouse commented:
Hi @cre8ivepark! The materials render correctly on my end, using 2018.4.12f. To double-check, I switched to another branch and back again; I then closed Unity, deleted the Library folder, and opened up the project again.
Materials appear correctly:
![image](https://user-images.githubusercontent.com/38055323/68523383-49bc7600-026d-11ea-9867-67c4bc22a960.png)
The pressable button prefab (Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/PressableButtonHoloLens2.prefab) also renders as expected (there are no edits to it):
![PressableButton2](https://user-images.githubusercontent.com/38055323/68523388-5771fb80-026d-11ea-83aa-fdab02ff3011.png)

I'm not sure why we're getting different results. I'll try cloning to my laptop to see if I can repro.

alandergrouse commented:
Ok! Just cloned the repo to my laptop and opened up the scene, and I'm not repro'ing. This is a fresh clone of mrtk with a fresh install of 2018.4.12f.

Here's what I see for the HandMenuLayoutExamples scene:
![image](https://user-images.githubusercontent.com/38055323/68523697-71154200-0271-11ea-9437-9540442cb359.png)

And the PressableButtonHoloLens2.prefab:
![image](https://user-images.githubusercontent.com/38055323/68523704-8c804d00-0271-11ea-830f-b8b6c85dac8f.png)

I think we're good!

Troy-Ferrell commented:
> Thanks for the feedback @Troy-Ferrell! Just pushed a commit with fixes for most of the issues you called out:
> 
> **Material instancing**
> Is this something specific to the current scene, or the project in general? FWIW, only one hand menu is enabled at a time – I'm not sure if that makes a difference though. Is there something I can do to optimize, without impacting the design?
> 
> **Prefab fixes**
> Good catch on the missing parameters in the prefabs. There are some overrides in the scene that I've applied to the prefabs. This fixed the missing shaders in the 3x4 prefab as well as the HandConstraintPalmUp solvers.
> 
> **Overlap solvers**
> The overlap solvers on the circular hand menu gives each button a different move lerp parameter, which makes for an "elastic" effect when the menu is moved along the hand. Here's a preview:
> ![GifMaker_20191108142358028](https://user-images.githubusercontent.com/38055323/68514778-8a979900-0233-11ea-8918-3035bbcc95fe.gif)
> 
> **Layout button selector**
> One activates the hand menu, and the other ensures that repeated pressed of the toggle button don't turn off the BackPlateToggleState node. I hear you on the scalability problem. Ideally, we could drop in an array of game objects and run actions on them at once; I'm not sure how to go about that though. Perhaps we can edit the Interactable script?

Material Instancing is not a problem specific to your scene. It can impact almost all applications. The scene in this PR is just particularly prone because you have so many icons/materials aka instances instead of re-using. This is an issue for icons on buttons in general for MRTK. I'm looking into this now.

Regarding the selector, the easy option is to create a script component that is instructed to activate a certain configuration. You could call the method on the script of what index or configuration is desired. or better yet, the script could just respond to OnSelection of the InteractableToggleCollection and based on the index active, use that to setup a configuration

Troy-Ferrell commented:
> OK, since there's currently no way to reduce number of assets. When @Troy-Ferrel makes progress on material instancing, perhaps we can revisit reducing the number of assets in this example.

The solution is this issue: #4660 . I am testing it out now by using [FontAwesome as an example](https://answers.unity.com/questions/1434980/how-to-add-font-awesome-into-unity-using-textmesh.html)

Troy-Ferrell commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Textures/PlaceholderIcon.png.meta:
```diff
@@ -0,0 +1,88 @@
+fileFormatVersion: 2
+guid: 29b54a82db8f0554986bdcdb1e53b4e9
```
Why do we need placeholdericon and placeholdericon_small?

alandergrouse replied:
Removed!

Troy-Ferrell commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Textures/gradient_1.png.meta:
```diff
@@ -0,0 +1,88 @@
+fileFormatVersion: 2
+guid: 80c1e1d0180de774980573f1aca2de4f
```
What are the gradients used for?

alandergrouse replied:
Good catch, I was previously using them to drive tile icon colors, but opted for just flat color instead. Removed.

Troy-Ferrell commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Prefabs/PressableButtonHoloLens2_HandMenu_Circular.prefab:
```diff
@@ -0,0 +1,1622 @@
+--- !u!114 &1171651852186010226
+MonoBehaviour:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
```
On CompressableCyldinersVisuals, there is a body, edge, and face gameobjects/meshes. 
Edge doesn't seem to really do anything that I can visually see when I toggle on/off
Body/Face also seem like they can be combined into one mesh with same material?

Troy-Ferrell replied:
Furthermore, how is this exactly different/needed compared to the prefabs "PressableButtonhololensCircular_nxn"?

alandergrouse replied:
I've replaced the buttons in the circular hand menu with PressableButtonHololensCircular prefaband added the "body" and "face" geometry to improve the the highlight cage (they use the same material); also, removed "edge" geo (and its material).

Troy-Ferrell commented on Assets/MixedRealityToolkit.Examples/Experimental/HandMenuLayout/Materials/HandMenu_ToggleDialog_Backplate.mat:
```diff
@@ -0,0 +1,191 @@
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_Name: HandMenu_ToggleDialog_Backplate
```
Why does the `HolographicBackPlate` material not work here?

alandergrouse replied:
Removed! I applied the backing material from the hand menus instead. The reasoning was that I wanted to tweak some of the material's parameters without breaking HoloGraphicBackPlate across the rest of the project.

cre8ivepark commented:
/azp run

davidkline-ms commented:
@Troy-Ferrell, can you revisit your review of this change and either approve or recommend we postpone to 2.3.0?

alandergrouse commented:
Thanks for helping guide this through, everyone! I really appreciate all your patience and feedback. ✌
